src/utp_stream.cpp

```diff
@@ -962,7 +962,10 @@ void utp_socket_impl::writable()
 	m_stalled = false;
 	if (should_delete()) return;
 
-	while(send_pkt());
+	// if the socket stalled while sending an ack then there will be a
+	// pending deferred ack. make sure it gets sent out
+	if (!m_deferred_ack || send_pkt(pkt_ack))
+		while(send_pkt());
```

Comment:
How about this: If there's no outstanding unsent ack then go straight to the `send_pkt()` loop like normal, otherwise use the `pkt_ack` flag for the first call, then proceed into the loop if needed. It's much simpler, and I think will achieve the same thing.

Reply:
that sounds good